### ISSUE TITLE
Finetune XLNet on STS-B with one V100 16GB GPU does not converge

### ISSUE THREAD

**Issue by cockroachzl** [zihangdai/xlnet#115]
Hi,

I try to reproduce the result of STS-B by following https://github.com/zihangdai/xlnet/blob/master/README.md#1-sts-b-sentence-pair-relevance-regression-with-gpus.

Since I only have access to a single V100 16GB GPU, so I changed the finetuning parameters to
```
CUDA_VISIBLE_DEVICES=0 python run_classifier.py \
  --do_train=True \
  --do_eval=False \
  --task_name=sts-b \
  --data_dir=${GLUE_DIR}/STS-B \
  --output_dir=proc_data/sts-b \
  --model_dir=exp/sts-b \
  --uncased=False \
  --spiece_model_file=${LARGE_DIR}/spiece.model \
  --model_config_path=${LARGE_DIR}/xlnet_config.json \
  --init_checkpoint=${LARGE_DIR}/xlnet_model.ckpt \
  --max_seq_length=128 \
  --train_batch_size=8 \
  --num_hosts=1 \
  --num_core_per_host=1 \
  --learning_rate=5e-5 \
  --train_steps=4800 \
  --warmup_steps=120 \
  --save_steps=600 \
  --is_regression=True
```
The changes I made are:
1. `CUDA_VISIBLE_DEVICES=0,1,2,3` to `CUDA_VISIBLE_DEVICES=0`
2. ` --num_core_per_host=4` to `--num_core_per_host=1`
3. `--train_steps=1200` to `--train_steps=4800` because now there is only one GPU

I observed the loss fluctuated between 1.x and 3.x. 

Evaluation gives me a pearson correlation of **0.2**, which seems suggest the model does not really learn anything significant.
```
I0704 00:08:11.761079 140663250270016 tf_logging.py:115] Eval result | eval_loss 2.539041757583618 | eval_pearsonr 0.08090152591466904 | global_step 1800 | loss 2.554678201675415 | path exp/sts-b/model.ckpt-1800 | step 1800 |
0704 00:08:45.479145 140663250270016 tf_logging.py:115] Eval result | eval_loss 2.354072332382202 | eval_pearsonr 0.05278123915195465 | global_step 2400 | loss 2.3670082092285156 | path exp/sts-b/model.ckpt-2400 | step 2400 |
I0704 00:09:19.798843 140663250270016 tf_logging.py:115] Eval result | eval_loss 2.393556833267212 | eval_pearsonr 0.03438245877623558 | global_step 3000 | loss 2.4071996212005615 | path exp/sts-b/model.ckpt-3000 | step 3000 |
I0704 00:10:13.699675 140663250270016 tf_logging.py:115] Eval result | eval_loss 2.424685478210449 | eval_pearsonr 0.07427685707807541 | global_step 3600 | loss 2.438821792602539 | path exp/sts-b/model.ckpt-3600 | step 3600 |
I0704 00:11:39.557689 140663250270016 tf_logging.py:115] Eval result | eval_loss 2.37259578704834 | eval_pearsonr 0.19783729314804077 | global_step 4800 | loss 2.3858773708343506 | path exp/sts-b/model.ckpt-4800 | step 4800 |
I0704 00:11:39.557799 140663250270016 tf_logging.py:115] ================================================================================
I0704 00:11:39.557908 140663250270016 tf_logging.py:115] Best result | eval_loss 2.2780275344848633 | eval_pearsonr 0.3380669057369232 | global_step 1200 | loss 2.2890121936798096 | path exp/sts-b/model.ckpt-1200 | step 1200 |
```

Could someone shed some light on what I have missed?

Thanks!

**Comment by cockroachzl** [comment]
I also tested the STS-B result in a AWS p3.8xlarge(4 V100 GPUs) using exactlyl the same parameter for finetuning and evaluation as https://github.com/zihangdai/xlnet#1-sts-b-sentence-pair-relevance-regression-with-gpus. 

Here are the evaluation result:
```
I0704 00:42:15.634207 140277864027904 run_classifier.py:786] ================================================================================
I0704 00:42:15.634326 140277864027904 run_classifier.py:790] Eval result | eval_loss 2.3103933334350586 | eval_pearsonr -0.006279603112488985 | global_step 1200 | loss 2.3223626613616943 | path exp/sts-b/model.ckpt-1200 | step 1200 |
I0704 00:42:15.634414 140277864027904 run_classifier.py:795] ================================================================================
I0704 00:42:15.634499 140277864027904 run_classifier.py:799] Best result | eval_loss 2.3722100257873535 | eval_pearsonr -0.0038009658455848694 | global_step 600 | loss 2.3854849338531494 | path exp/sts-b/model.ckpt-600 | step 600 |
```
The `eval_pearsonr` is close to 0, instead of `# Expected performance: "eval_pearsonr 0.916+ "`.

**Comment by huseinzol05** [comment]
Same!

**Comment by cockroachzl** [comment]
Just realized what I went wrong. I used python3 instead of python2. After switching to python2, I can get the claimed result.

```
I0704 01:11:24.028214 139952961881856 run_classifier.py:786] ================================================================================
I0704 01:11:24.028338 139952961881856 run_classifier.py:790] Eval result | eval_loss 0.394300609827 | eval_pearsonr 0.911117553711 | global_step 1200 | loss 0.445616304874 | path exp/sts-b/model.ckpt-1200 | step 1200 |
I0704 01:11:24.028434 139952961881856 run_classifier.py:795] ================================================================================
I0704 01:11:24.028522 139952961881856 run_classifier.py:799] Best result | eval_loss 0.394300609827 | eval_pearsonr 0.911117553711 | global_step 1200 | loss 0.445616304874 | path exp/sts-b/model.ckpt-1200 | step 1200 |
```
**eval_pearsonr 0.911117553711**

**Comment by huseinzol05** [comment]
What?! That is very weird. It just matter of syntax.

**Comment by cockroachzl** [comment]
@huseinzol05 Yeah, I feel it is weird as well! The only thing I changed is rather than using `source activate tensorflow_p36` to activate the python3 env, `source activate tensorflow_p27` is used instead and get the claimed result. https://docs.aws.amazon.com/dlami/latest/devguide/tutorial-tensorflow.html

**Comment by cockroachzl** [comment]
I think I closed this issue too early. I finetuned STS-B again in python3 with tensorflow 1.13.1 and this time it converged:
```
I0704 06:09:41.269094 140499994494720 estimator.py:2039] Saving 'checkpoint_path' summary for global step 1200: exp/sts-b/model.ckpt-1200
I0704 06:09:41.269527 140499994494720 run_classifier.py:794] ================================================================================
I0704 06:09:41.269653 140499994494720 run_classifier.py:798] Eval result | eval_loss 0.39375582337379456 | eval_pearsonr 0.9153165817260742 | global_step 1200 | loss 0.3939589560031891 | path exp/sts-b/model.ckpt-1200 | step 1200 |
I0704 06:09:41.269742 140499994494720 run_classifier.py:803] ================================================================================
I0704 06:09:41.269829 140499994494720 run_classifier.py:807] Best result | eval_loss 0.39375582337379456 | eval_pearsonr 0.9153165817260742 | global_step 1200 | loss 0.3939589560031891 | path exp/sts-b/model.ckpt-1200 | step 1200 |
(tensorflow_p36) ubuntu@ip-172-31-88-214:~/xlnet$
```
**eval_pearsonr 0.9153165817260742**

Therefore, I think python version may not be the cause of the convergence issue. Right now I doubted if the issue is because of the random initialization. Maybe for certain randomly initialized parameters, the trainer could not converge.

Again, the above result is obtained in a p3.8xlarge ec2 host with 4  V100 16GB GPUs.

**Comment by cockroachzl** [comment]
Another observation is, I fine tuned SST-2 twice, each with 10k steps in python 2 with tensorflow 1.13.1 in p3.8xlarge. The first attempt failed to converge and the second attempt succeeded with about the same result in the paper. These two runs indeed had some training parameter differences:
1. In the first run, the learn rate is set to 5e-5 whereas it is set to 2e-5 in the second run
2. The warmup steps is set to 120 in the first run and 1k in the second run.
The second run used parameters suggested in https://github.com/zihangdai/xlnet/issues/96#issuecomment-507359322

**Comment by huseinzol05** [comment]
Can you try to reduce dropout to 0?

**Comment by alexpython1988** [comment]
I repeated the STS-B on one NVIDIA Quadro P6000 with train step set to 5400 and learning rate set to 1e-5. I could get the  eval_pearsonr to 0.9157, which is very close to the claim in the repo.

**Comment by Qi-HuT** [comment]
![image](https://user-images.githubusercontent.com/46215418/88993836-970bd680-d319-11ea-93a1-438fad5c5bf8.png)
the 
The author explains that python2 is used
